### ISSUE TITLE
iPhoto on 12.3 and 12.3.1

### ISSUE THREAD

**Issue by jb59** [cormiertyshawn895/Retroactive#280]
Looks like 12.3 may have killed iPhoto? - instant crash with  "iPhoto quit unexpectedly" message

Crash report begins:
_________________________________________________________
Process:               iPhoto [677]
Path:                  /Applications/iPhoto.app/Contents/MacOS/iPhoto
Identifier:            com.apple.iPhoto9
Version:               9.6.1 (910.42)
Build Info:            iPhotoProject-910042000000000~1
Code Type:             X86-64 (Native)
Parent Process:        launchd [1]
User ID:               501

Date/Time:             2022-03-24 16:09:20.9326 +0000
OS Version:            macOS 12.3 (21E230)
Report Version:        12
Anonymous UUID:        6AE9D0FB-BDCF-1A30-F2B8-88A1D34BECBF

Time Awake Since Boot: 240 seconds

System Integrity Protection: enabled

Crashed Thread:        0

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    Namespace DYLD, Code 1 Library missing
Library not loaded: /System/Library/Frameworks/Python.framework/Versions/2.6/Python
Referenced from: /Applications/iPhoto.app/Contents/MacOS/iPhoto
Reason: tried: '/System/Library/Frameworks/Python.framework/Versions/2.6/Python' (no such file), '/Library/Frameworks/Python.framework/Versions/2.6/Python' (no such file)
(terminated at launch; ignore backtrace)
 ____________________________________________________________________________________

Weirdly perhaps, Aperture still running as before!

**Comment by jb59** [comment]
Update on this - just read an article on MacRumors about 12.3 removing Python 2  - suspect this may be the issue.

**Comment by jb59** [comment]
Update and solution - 
Disabled SIPS via terminal in Recovery Mode
copied Python.framework folder from pre Monterey 12.3 Time Machine backup into /Library/Frameworks folder
( folder contains Python versions 2.3, 2.5, 2.6. 2.7)   
Enabled SIPS via terminal in Recovery Mode
IPhoto runs as before

**Comment by jb59** [comment]
Thanks again for Retroactive!

**Comment by jb59** [comment]
Update  - Monterey 12.3.1 update causes the issue to reappear - instant iPhoto crash with Python 2.6 library issue.

Solution:
Copy Python.framework folder from pre-Monterey 12.3 Time Machine backup to:
/Library/Frameworks

iPhoto runs as before (Python library verification message on first iPhoto start)

**Note: no need to enable/disable SIPS**

**Comment by doxxxob** [comment]
This solution don't work for me. There is no Python.framework at /Library/Framework folder now and back in 2021.
Confusion. Any suggestions?
Thank for Retroactive which work so well since the Monterey update.

**Comment by jb59** [comment]
Ok - have checked on my MacBook Air, running BigSur 11.6.5 - location of Python framework is different to the above and is at:

/System/Library/Frameworks 

If you have a TM backup it should be there (assuming you were running iPhoto at that time)

**Comment by drrules** [comment]
Hey JB59, thanks for the info.

Unfortunately, it won't let me copy the files to the system folder (even with SIPS off or with a Root account or anything else I can think of).
I tried running a "Migrate from Time Machine" and it won't let me select which files I want to copy over. Tried just opening TM and it won't let me copy that too. Should I just restore my "Settings and Other Files" is that what you did?

I tried putting it into ~/Library/Frameworks and it just won't launch iPhoto... seems like it's gotta be /System/Library/Frameworks and I can't write to it.

Any ideas?
Thanks.

**Comment by jb59** [comment]
Have been messaged about this, so here are a couple of screen grabs showing my (iPhoto functioning) Python Framework 

<img width="1250" alt="Screenshot 2022-04-07 at 08 08 37" src="https://user-images.githubusercontent.com/29352388/162141760-cb3414d7-a813-44a6-8e9a-08b4681b1843.png">
<img width="1249" alt="Screenshot 2022-04-07 at 08 09 05" src="https://user-images.githubusercontent.com/29352388/162141840-908909e8-52d8-48e4-baa3-38038fd302ce.png">

It can be seen that "versions" 2.3, 2.5, 2.6 and Current are all aliases pointing to the 2.7 folder - the original crash report quoted above shows that iPhoto is looking for Python 2.6 , but appears to be happy to be "redirected" to 2.7.  

The Python.framework folder on my system (2015 5k iMac) is reported as 208MB - it compresses to 87.8mb so if anyone wants to try it on their system, message me and I'll put it up on a fileshare platform.

**Comment by jb59** [comment]
Hi There,

have replied on Github thread - hope it helps!

Cheers,
JB59

________________________________

**Comment by thisplantguy** [comment]
Hey there-- have an M1 MacBook Air and have been using Time Machine since last July.  I tried looking for the python framework folder in both locations, but TM seems to have chosen not to back that folder up.  I'd love to get iPhoto running again (I'm on MacOS 12.3.1).  Would you be willing to fileshare the compressed file?  Thanks!!

Update/ edit: I found the folder on an older clone of my previous MacBook Air, however the OS won't let me copy anything into my /System folder.

**Comment by Tetonne** [comment]
can someone share the missing folder please?

**Comment by RedTartan04** [comment]
@thisplantguy It's supposed to go into /Library/Frameworks not /System/Library/Frameworks. /System is write protected.

**Comment by RedTartan04** [comment]
@jb59 Thanks a ton. I copied it from a BigSur backup to  /Library/Frameworks as per your screenshots. Still crashing with this message
'/Library/Frameworks/Python.framework/Versions/2.6/Python' (code signature invalid (errno=1)
:-(

**Comment by doxxxob** [comment]
Is python realy the problem? I don't belief it. Did anyone had success with this??

**Comment by RedTartan04** [comment]
> Is python realy the problem? I don't belief it. Did anyone had success with this??

Why don't you believe it? OP has posted screenshots. Also if you look into the crash message, it's clearly looking for python and not getting it.

```
Termination Reason:    Namespace DYLD, Code 1 Library missing
Library not loaded: 
/System/Library/Frameworks/Python.framework/Versions/2.6/Python
Referenced from: /Applications/iPhoto.app/Contents/MacOS/iPhoto
Reason: tried: '/System/Library/Frameworks/Python.framework/Versions/2.6/Python' (no such file), '/Library/Frameworks/Python.framework/Versions/2.6/Python' (code signature invalid (errno=1) sliceOffset=0x00004000, codeBlobOffset=0x00168390, codeBlobSize=0x000041F0 for '/Library/Frameworks/Python.framework/Versions/2.7/Python')
(terminated at launch; ignore backtrace)
```

**Comment by RedTartan04** [comment]
It's working now :-) Did a manual code sign:
`sudo codesign -f -s - /Library/Frameworks/Python.framework/Versions/2.7`

Before that I also did a `sudo chown -R root /Library/Frameworks/Python.framework` after I copied it over from a BigSur backups /System/Library/Frameworks. Not sure if that's needed though.

**Comment by drrules** [comment]
Asking someone to share a folder with you might not be the wisest thing. 

Anyhow, try this: https://www.python.org/downloads/release/python-2718/

Once it finishes installing you'll have the python folder with 2.7 in there (in /Library/Frameworks, as shown above. We can't write to /System/Library anymore.) Then create the aliases manually for 2.3-2.6 and current. That should fix your problem.

Thank you JB59 for the assistance!!

**Comment by RedTartan04** [comment]
For those who wont believe until seeing ;-)
<img width="400" alt="iPhoto_at_Monterey_csw" src="https://user-images.githubusercontent.com/99086095/162574854-baa82ec2-4d05-45cf-963d-c70d270f68b4.png">

**Comment by Tetonne** [comment]
just install https://www.python.org/ftp/python/2.7.18/python-2.7.18-macosx10.9.pkg on 12.3.1
then sudo codesign -f -s - /Library/Frameworks/Python.framework/Versions/2.7
created alias of 2,7 for 2,3 ; 2,5 ; 2,6 (drag and option+cmd)
but iPhoto still doesn't work :-(

**Comment by doxxxob** [comment]
Also making it the drrules-way. Does't work

**Comment by doxxxob** [comment]
My crash message: 

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       UNKNOWN_0xD at 0x0000000000000000
Exception Codes:       0x000000000000000d, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [6330]

**Comment by Tetonne** [comment]
Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    Namespace DYLD, Code 1 Library missing
Library not loaded: /System/Library/Frameworks/Python.framework/Versions/2.6/Python
Referenced from: /Applications/iPhoto.app/Contents/MacOS/iPhoto
Reason: tried: '/System/Library/Frameworks/Python.framework/Versions/2.6/Python' (no such file), '/Library/Frameworks/Python.framework/Versions/2.6/Python' (no such file)
(terminated at launch; ignore backtrace)

Application Specific Information:
Library not loaded: /System/Library/Frameworks/Python.framework/Versions/2.6/Python
Referenced from: /Applications/iPhoto.app/Contents/MacOS/iPhoto
Reason: tried: '/System/Library/Frameworks/Python.framework/Versions/2.6/Python' (no such file), '/Library/Frameworks/Python.framework/Versions/2.6/Python' (no such file)

**Comment by RedTartan04** [comment]
@Tetonne If Finder Alias wont work you may need a symlink in Terminal instead (although I'm still not sure if it's relevant to `chown` everything to root - anything you do with Finder will have your own user as owner, not root).
```
cd /Library/Frameworks/Python.framework/Versions
ln -s 2.7 2.6
```
To check: `ls -la` should look like
```
lrwxr-xr-x   1 root  wheel    3  1 Jan  2020 2.3 -> 2.7
lrwxr-xr-x   1 root  wheel    3  1 Jan  2020 2.5 -> 2.7
lrwxr-xr-x   1 root  wheel    3  1 Jan  2020 2.6 -> 2.7
drwxr-xr-x  12 root  wheel  384  9 Apr 14:43 2.7
lrwxr-xr-x   1 root  wheel    3  1 Jan  2020 Current -> 2.7
```
while a Finder Alias looks like
`-rwxr-xr-x@  1 frank  wheel  768  9 Apr 20:18 2.7 Alias
`

**Comment by RedTartan04** [comment]
> My crash message:
> 
> Crashed Thread: 0 Dispatch queue: com.apple.main-thread
> 
> Exception Type: EXC_BAD_ACCESS (SIGSEGV) Exception Codes: UNKNOWN_0xD at 0x0000000000000000 Exception Codes: 0x000000000000000d, 0x0000000000000000 Exception Note: EXC_CORPSE_NOTIFY
> 
> Termination Reason: Namespace SIGNAL, Code 11 Segmentation fault: 11 Terminating Process: exc handler [6330]

Different reason... Did you try to rerun Retroactive and force it to enable iPhoto again?
Not sure if related, but it's someting I did before fixing Python.

**Comment by thisplantguy** [comment]
I want to use iPhoto as much as the next person in this thread, but playing around with root permissions and making system aliases seem to be a bit of a stretch, security-wise, to do this.  I do hope that the dev can come up with a more elegant solution for this, but as it stands now, I'm not comfortable playing around with signing packages.

**Comment by Tetonne** [comment]
**How to I did and it works** 
thanks to [jb59](https://github.com/jb59) ; [RedTartan04](https://github.com/RedTartan04) ; [drrules](https://github.com/drrules) 

1. install https://www.python.org/ftp/python/2.7.18/python-2.7.18-macosx10.9.pkg

2.  In Terminal : sudo codesign -f -s - /Library/Frameworks/Python.framework/Versions/2.7

3. Create the needed alias :
cd /Library/Frameworks/Python.framework/Versions
sudo ln -s 2.7 2.6
sudo ln -s 2.7 2.5
sudo ln -s 2.7 2.3

huge thanks

**Comment by RedTartan04** [comment]
> I want to use iPhoto as much as the next person in this thread, but playing around with root permissions and making system aliases seem to be a bit of a stretch, security-wise, to do this. I do hope that the dev can come up with a more elegant solution for this, but as it stands now, I'm not comfortable playing around with signing packages.

Relax ;-) There's nothing spooky about this installation process. It doesn't work around SIP for example. Every step is pretty normal stuff, if you're a developer or admin.

In my case I took the package from a backup of my own mac, so signing it myself is perfectly ok. Also with the official package I don't see a problem to use and sign that.

We're not 'playing with root permissions'. I did a chown, but that doesn't seem to be required. Even if it was, it just restores things to what they were originally before 12.3. 

Also there are no 'system' aliases. Same as above: restoring things to what they were. Again, just very basic unix stuff.

A 'more elegant' solution would also have to install Python one way or another. Not sure if iPhoto could be patched to look at a different location, e.g. in a user's directory or if Python could be put into the app bundle. Either way it would involve the install package, code signing and probably a symlink or two ;-) It's just how things work.

**Comment by doxxxob** [comment]
@RedTartan04 

    My crash message:

    Crashed Thread: 0 Dispatch queue: com.apple.main-thread

    Exception Type: EXC_BAD_ACCESS (SIGSEGV) Exception Codes: UNKNOWN_0xD at 0x0000000000000000 Exception Codes: 0x000000000000000d, 0x0000000000000000 Exception Note: EXC_CORPSE_NOTIFY

    Termination Reason: Namespace SIGNAL, Code 11 Segmentation fault: 11 Terminating Process: exc handler [6330]

Different reason... Did you try to rerun Retroactive and force it to enable iPhoto again?
Not sure if related, but it's someting I did before fixing Python.


After rerun Retroactive: same crash.

**Comment by pdq2** [comment]
Just worked for me (see Tetonne, above). Monterey 12.3.1, on M1 Mac mini. Still some crashes, but certainly better than crashing on open, as before.

**Comment by drrules** [comment]
Hey everyone, just wanted to confirm it's working on Monterey 12.5 too. Just upgraded and iPhoto broke.
Popped open terminal and reran the codesign from @Tetonne:

sudo codesign -f -s - /Library/Frameworks/Python.framework/Versions/2.7

All the aliases remained so that wasn't an issue.

If it fails for you again, just re-install python from my link above again.

Gotta love this community!

**Comment by doxxxob** [comment]
Don’t work for me. iPhoto didn’t  start at all. I am depressed.

**Comment by drrules** [comment]
> Don’t work for me. iPhoto didn’t start at all. I am depressed.
> […](#)
> Am 30.Juli.2022 um 13:49 schrieb drrules ***@***.***>: Hey everyone, just wanted to confirm it's working on Monterey 12.5 too. Just upgraded and iPhoto broke. Popped open terminal and reran the codesign from @Tetonne: sudo codesign -f -s - /Library/Frameworks/Python.framework/Versions/2.7 All the aliases remained so that wasn't an issue. If it fails for you again, just re-install python from my link above again. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented.

Have you followed all the instructions above?

First install python from my source link or @Tetonne 's which is a direct link to the download.
Then make the symlinks thanks to the suggestion from @RedTartan04 and following the instructions for 2.7 down to 2.3 that @Tetonne mentions.
Run the Retroactive app. 
Then codesign python like in my previous post.

Also, maybe delete iPhoto and all of its related files and grab the installer from archive.org (just search for iphoto 9.6.1 archive on google).

**Comment by doxxxob** [comment]
I installed Python 2.7, run Retroctive again and code designed  Python. But my iPhoto want’s

Termination Reason:    Namespace DYLD, Code 1 Library missing
Library not loaded: /System/Library/Frameworks/Python.framework/Versions/2.6/Python
Referenced from: /Applications/iPhoto.app/Contents/MacOS/iPhoto
Reason: tried: '/System/Library/Frameworks/Python.framework/Versions/2.6/Python' (no such file), '/Library/Frameworks/Python.framework/Versions/2.6/Python' (no such file)
(terminated at launch; ignore backtrace)

There I cannot find the Python.framework

My Python.framework is at /Library/Frameworks NOT at /System/Library/Frameworks

In the System folder I cannot change anything. 

What are symlinks? And how to proceed?

**Comment by drrules** [comment]
Go to your Utilities folder and open up terminal.
First paste this in: sudo codesign -f -s - /Library/Frameworks/Python.framework/Versions/2.7
It'll ask you for your admin password, type that in.

Then paste in this:
cd /Library/Frameworks/Python.framework/Versions
hit enter
Then paste in this:
sudo ln -s 2.7 2.6
hit enter

Try that.

**Comment by thisplantguy** [comment]
Hey there— 

So I was able to do the code sign with no problem, but after I get into the Versions directory in Terminal and use your ln -s command, I get the following:

% sudo ln - s 2.7 2.6
ln: 2.6: No such file or directory


And after that, iPhoto still crashes upon trying to open.

**Comment by thisplantguy** [comment]
Ah sorry, my bad… caught my own error.  I inserted a space after the - and before “s”.  It’s working now!!

**Comment by drrules** [comment]
There you go! Huzzah!

**Comment by doxxxob** [comment]
First of all. 
Thank you for the help.
I made it in this order.
After 
sudo ln -s 2.7 2.6
the answer was 
ln: 2.6: File exists
after this iPhoto crushes with the same:

Termination Reason:    Namespace DYLD, Code 1 Library missing
Library not loaded: /System/Library/Frameworks/Python.framework/Versions/2.6/Python
Referenced from: /Applications/iPhoto.app/Contents/MacOS/iPhoto
Reason: tried: '/System/Library/Frameworks/Python.framework/Versions/2.6/Python' (no such file), '/Library/Frameworks/Python.framework/Versions/2.6/Python' (no such file)
(terminated at launch; ignore backtrace)

Again my feeling is that everything is OK in /Library/Frameworks but nothing is in /System/Library/Frameworks
I couldn’t change anything in /System.
Again Thank you

**Comment by drrules** [comment]
@doxxxob I had that error before too. It was fixed by the code signature. I also installed a few different versions of Python.

**Comment by RedTartan04** [comment]
> Again my feeling is that everything is OK in /Library/Frameworks but nothing is in /System/Library/Frameworks I couldn’t change anything in /System. Again Thank you
> […](#)

Again, your feeling is wrong @doxxxob ;-) As the error message says, it first tries /System/Library/Frameworks/... and then /Library/Frameworks/.... Only _one_ is needed, not both, and we install/make links in /Library/Frameworks/... to satisfy this search.

Please do the following and post the output, so we may help you:
`cd /Library/Frameworks/Python.framework/Versions`
`ls -la`

**Comment by drrules** [comment]
> Again, your feeling is wrong @doxxxob ;-) As the error message says, it first tries /System/Library/Frameworks/... and then /Library/Frameworks/.... Only _one_ is needed, not both, and we install/make links in /Library/Frameworks/... to satisfy this search.
> 
> Please do the following and post the output, so we may help you: `cd /Library/Frameworks/Python.framework/Versions` `ls -la`

Good call @RedTartan04:
@doxxxob This had me rereading the error post. You can also try deleting all the folders/aliases for everything not 2.7 and following our instructions above to create the aliases again. 



Also just a general question to my fellow techies here, has anyone tried this on the Ventura beta yet?

**Comment by doxxxob** [comment]
First when I put cd /Library/Frameworks/Python.framework/Versions ls -la in the terminal the answer is:
cd: too many arguments

Second when go to the directory Versions by hand and type ls the answer is:
2.3 	2.5	2.6	2.6 a	2.7	3.1	3.2	3.3	Current

Thirdly when I type in Versions ls -la the answer is:
total 24
drwxr-xr-x  11 root          wheel  352 30 Jul 15:59 .
drwxr-xr-x@  6 root          wheel  192 30 Jul 15:59 ..
-rwxr-xr-x@  1 thomashornig  wheel  768  9 Apr 16:43 2.3 
-rwxr-xr-x@  1 thomashornig  wheel  768  9 Apr 16:47 2.5
-rwxr-xr-x@  1 thomashornig  wheel  768  9 Apr 16:48 2.6
drwxrwxr-x  10 root          wheel  320 31 Aug  2010 2.6 a
drwxrwxr-x  12 root          admin  384 30 Jul 17:34 2.7
drwxrwxr-x   9 root          wheel  288 24 Mär  2010 3.1
drwxrwxr-x   9 root          wheel  288 10 Apr  2012 3.2
drwxrwxr-x   9 root          wheel  288  6 Apr  2013 3.3
lrwxr-xr-x   1 root          wheel    3 30 Jul 15:59 Current -> 2.7

Did this help?
I hope so.
Thank you very much for your effort.

**Comment by RedTartan04** [comment]
oh dear, you have to follow the instructions more precisely.
cd and ls are two commands, not one :)

I‘m not on the Mac right now but this
> -rwxr-xr-x@ 1 thomashornig wheel 768 9 Apr 16:48 2.6
doesn‘t look like the needed symlink? Delete it and create the symlink, using the ls command, as explained many times above (sometime called alias, but an alias is a different kind of link and wont work, as explained…)

read carefully, execute exactly as told; no room for error.

**Comment by RedTartan04** [comment]
sorry meant ln command not ls, but you have to look it up above anyway

**Comment by drrules** [comment]
Hi all,
Just a reminder, if you update your operating system, you will need to run the codesign again.

sudo codesign -f -s - /Library/Frameworks/Python.framework/Versions/2.7

I just updated to 12.5.1 and I can confirm this process is still working.

**Comment by doxxxob** [comment]
It worked after changing the alias in symlink as described above.
Thanks a lot RedTartan 04 and drrules

**Comment by cormiertyshawn895** [comment]
Thanks everyone for chiming in! This is fixed in [Retroactive 2.0](https://github.com/cormiertyshawn895/Retroactive/releases/tag/2.0).
